cocos/animation/marionette/graph-eval.ts

```diff
@@ -200,42 +210,38 @@ export class AnimationGraphEval {
     }
 
     public getLayerWeight (layerIndex: number) {
-        assertIsTrue(layerIndex >= 0 && layerIndex < this._layerEvaluations.length, `Invalid layer index`);
-        return this._layerEvaluations[layerIndex].weight;
+        return this._rootPoseNode.getLayerWeight(layerIndex);
     }
 
     public setLayerWeight (layerIndex: number, weight: number) {
-        assertIsTrue(layerIndex >= 0 && layerIndex < this._layerEvaluations.length, `Invalid layer index`);
-        this._layerEvaluations[layerIndex].weight = weight;
+        this._rootPoseNode.setLayerWeight(layerIndex, weight);
     }
 
     public overrideClips (overrides: ReadonlyClipOverrideMap) {
         const {
             _poseLayoutMaintainer: poseLayoutMaintainer,
-            _layerEvaluations: layerEvaluations,
         } = this;
 
         poseLayoutMaintainer.startBind();
 
-        const nLayers = layerEvaluations.length;
-        for (let iLayer = 0; iLayer < nLayers; ++iLayer) {
-            const layerEval = layerEvaluations[iLayer];
-            layerEval.overrideClips(overrides, this._bindingContext);
-        }
+        this._rootPoseNode.overrideClips(overrides, this._bindingContext);
 
         this._updateAfterPossiblePoseLayoutChange();
     }
 
+    private _rootPoseNode: DefaultTopLevelPoseNode;
     private _varInstances: Record<string, VarInstance> = {};
     private _hasAutoTrigger = false;
     private _metaValueRegistry = new MetaValueRegistry();
     private _poseLayoutMaintainer: AnimationGraphPoseLayoutMaintainer;
     private _bindingContext: AnimationGraphBindingContext;
+    private _settleContext: AnimationGraphSettleContext;
     /**
      * Preserved here for clip overriding.
      */
     private declare _root: Node;
     private declare _evaluationContext: AnimationGraphEvaluationContext;
+    private _rootUpdateContextGenerator = new AnimationGraphUpdateContextGenerator();
```

Comment:
Why don't create `updateContext` directly?

Reply:
Change the value directly?

```
function updateTheStateMachine(context: UpdateContext) {
  this._updateContext.setDeltaTime(context.deltaTime * 1.0);
  this.stateA.update(this._updateContext);
  this._updateContext.setDeltaTime(context.deltaTime * 0.7);
  this.stateB.update(this._updateContext);
}
```

Reply:
If so, you're missing the `indicativeWeight` field:

```diff
 function updateTheStateMachine(context: UpdateContext) {
   this._updateContext.setDeltaTime(context.deltaTime * 1.0);
+ this._updateContext.setIndicativeWeight(context.indicativeWeight * stateAWeight);
   this.stateA.update(this._updateContext);
   this._updateContext.setDeltaTime(context.deltaTime * 0.7);
+ this._updateContext.setIndicativeWeight(context.indicativeWeight * stateBWeight);
   this.stateB.update(this._updateContext);
}
```

I think it's equivalent. My design is "update context is **immutable**".